pyat/at/acceptance/touschek.py

```diff
@@ -0,0 +1,215 @@
+import at
+import numpy
+from ..lattice import Lattice, AtError
+from scipy.special import iv
+from scipy import integrate
+from scipy.optimize import fsolve
+from ..lattice.constants import clight, e_mass, qe, clight, _e_radius
+from .acceptance import get_momentum_acceptance
+
+
+__all__ = ['get_bunch_length_espread', 'get_lifetime']
+
+
+def get_bunch_length_espread(ring, zn=None, bunch_curr=None, espread=None):
+    """
+    Solves the Haissinski formula and returns the bunch length and energy
+    spread for given bunch current and Z/n. If both zn and bunch_curr are None,
+    zero current case, otherwise both are needed for the calculation
+
+    PARAMETERS
+        ring              ring use for tracking
+
+    KEYWORDS
+        zn=None           Z/n for the full ring
+        bunch_curr=None   Bunch current
+        espread=None      Energy spread, if None use lattice parameter
+
+    OUTPUT
+        Bunch length, energy spread
+    """
+    def haissinski(x, cst):
+        return x**3-x-cst
+
+    ep = at.envelope_parameters(ring.radiation_on(copy=True))
+    bl0 = ep.sigma_l
+    if espread is None:
+        espread = ep.sigma_e
+    if zn is None and bunch_curr is None:
+        bl = bl0
+    elif zn is None or bunch_curr is None:
+        raise AtError('Please provide both current and Z/n for bunch '
+                      'length calculation')
+    else:
+        vrf = ring.get_rf_voltage()
+        h = ring.harmonic_number
+        etac = ring.get_slip_factor()
```

Comment:
If I give a ring with radiation on, and a bunch current, and a zn to get_lifetime, i return an error here because ring.get_slip_factor needs rad off.

Reply:
replace with         `etac = at.get_slip_factor(ring.radiation_off(copy=True))`